/* XPECT_SETUP eu.numberfour.n4js.bugreports.tests.N4JSBugreportTest END_SETUP  */

/* XPECT output --- 
<==
stdout:
e.message === some message: true
e.notMsg === custom message: true
e.n === 36: true
e.method() === 36: true
e instanceof Error: true
e instanceof CustomError_007: true
e instanceof SubCustomError_007: true
e has stack trace: true
e.stack starts with 'Error: some message': true
stderr:
==>
--- */

class CustomError_007 extends Error {
	notMsg : string;
	constructor(notMsg : string, someMsg : string) {
		super(someMsg);
		this.notMsg = notMsg;
	}
}

class SubCustomError_007 extends CustomError_007 {
	n : number;
	constructor(anotherNotMsg : string, someMsg : string, n : number) {
		super(anotherNotMsg, someMsg);
		this.n = n;
	}
	method() : number {
		return this.n;
	}
}

var e = new SubCustomError_007('custom message', 'some message', 36);
console.log('e.message === some message:', e.message === 'some message');
console.log('e.notMsg === custom message:', e.notMsg === 'custom message');
console.log('e.n === 36:', e.n === 36);
console.log('e.method() === 36:', e.method() === 36);
console.log('e instanceof Error:', e instanceof Error);
console.log('e instanceof CustomError_007:', e instanceof CustomError_007);
console.log('e instanceof SubCustomError_007:', e instanceof SubCustomError_007);
// XPECT warnings --> "Unnecessary cast from SubCustomError_007 to Object" at "e as Object"
var stack = (e as Object)['stack'];
console.log('e has stack trace:', undefined !== stack);
console.log('e.stack starts with \'Error: some message\':', 0 === (stack as String).indexOf('Error: some message'));
